### PR TITLE
Implement HandshakeComplete message

### PR DESCRIPTION
**Pull request by sergeuz**
### Problem

At present, there is no explicit indication that the application handshake has completed. The cloud does not know with any certainty that the device has completed the application handshake and considers itself online. The possibility of out of order delivery of events compounds this problem even further.

### Solution

This PR changes the handshake procedure so that it looks as follows:

1. the application handshake is performed and various messages are sent to the cloud
2. the system waits for these messages to be acknowledged
3. once all messages are acknowledged,  a new `HandshakeComplete` CoAP message is sent to the cloud.
4. on acknowledgement of that message, `Particle.connected()` returns true.

### Steps to Test

_TBD_

### References

- [ch40066]

### CODE REVIEW

File: system/src/system_task.cpp [hunk 1]
```diff
@@ -347,79 +387,100 @@ int cloud_handshake()
...
+                    g_cloudHandshakeState = CloudHandshakeState::DONE;
+                }
+            } else if (err == protocol::ProtocolError::NOT_IMPLEMENTED) { // Core?
+                g_cloudHandshakeState = CloudHandshakeState::DONE;
+                err = 0;
+            }
+        }
+        if (g_cloudHandshakeState == CloudHandshakeState::DONE) {
+            // TODO: There's no protocol API to get the current session state, so we're running
+            // one more iteration of the communication loop to make sure all handshake messages
```
Comment (m-mcgowan) [inline system/src/system_task.cpp]: I am kind of puzzled by this one too!

File: communication/src/protocol.cpp [hunk 2]
```diff
@@ -292,8 +292,8 @@ int Protocol::begin()
 	ack_handlers.clear();
 	last_ack_handlers_update = callbacks.millis();
 
-	uint32_t channel_flags = 0;
-	ProtocolError error = channel.establish(channel_flags, application_state_checksum());
+	flags &= ~SKIP_SESSION_RESUME_HELLO;
```
[the remark this answers is not in the document]
Comment (sergeuz) [reply communication/src/protocol.cpp]: This fixes a bug where the `SKIP_SESSION_RESUME_HELLO` flag was persisting through multiple connection attempts.

File: system/src/system_task.cpp (hunk 1 above)
Comment (sergeuz) [reply system/src/system_task.cpp]: This is not a new code, but, yeah, the comment could be more clear. The problem is that when a CoAP message expires, the event loop function doesn't return an error (instead, the protocol implementation [closes the channel](https://github.com/particle-iot/device-os/blob/2a8f892646022f50f9e5572a8132811961ed6111/communication/src/coap_channel.cpp#L57-L63) silently), and we have to invoke the loop function one more time to determine if the current protocol session is still valid.

**Comment by sergeuz** [comment]
I'm closing this PR as it's unlikely for it to land in any Device OS release in its current form in foreseeable future.

**Pull request closed without merging**